vertical-pod-autoscaler/pkg/utils/status/status_object.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"net"
+	"time"
+
+	apicoordinationv1 "k8s.io/api/coordination/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	typedcoordinationv1 "k8s.io/client-go/kubernetes/typed/coordination/v1"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	// Parameters for retrying with exponential backoff.
+	retryBackoffInitialDuration = 100 * time.Millisecond
+	retryBackoffFactor          = 3
+	retryBackoffJitter          = 0
+	retryBackoffSteps           = 6
+)
+
+// Client for the status object.
+type Client struct {
+	client               typedcoordinationv1.LeaseInterface
+	name                 string
+	namespace            string
+	leaseDurationSeconds int32
+}
+
+// NewClient returns a client for the status object.
+func NewClient(c clientset.Interface, name, namespace string, leaseDuration time.Duration) *Client {
+	return &Client{
+		client:               c.CoordinationV1().Leases(namespace),
+		name:                 name,
+		namespace:            namespace,
+		leaseDurationSeconds: int32(leaseDuration) / int32(time.Second),
+	}
+}
+
+// UpdateStatus renews status object lease.
+// Status object will be created if it doesn't exist.
+func (c *Client) UpdateStatus() error {
```

Comment:
How often will this be called in practice?

Reply:
I want it to use for Admission Controller. I want to call `UpdateStatus` every 10s. The same practice is used for node leases.
Assuming that `UpdateStatus` correlates with number of master, the generated qps should be negligible.

I also used explicit get for retrieving status instead of using watch due to correctness reasons. We are not generating heavy load on master.

Reply:
What correctness will we sacrifice with watch? 
We might try to create an already existing object, which will just cause the create to fail. (This may happen anyway if someone creates lease between our get and create)
For Updating we handle conflicts anyway in line 74.
Am I missing something?

I agree this might not be a huge load on master, but we are tight on out qps to API server anyway, so I wan't to make sure we're not making things worse.

Reply:
Watch can lag. Explicit get guaranties up-to-date data.
My assumption is that explicit get is always better than watch. Imho watch should be used if this could potentially reduce load on apiserver. This is not our case here.
Plus, explicit get is easier to implement and to understand than watch.

Implement explicit get and then if it necessary changing to watch is also an option ;)
Btw, I can reach to scalability folks and ask about their option.

Reply:
I cannot agree that explicit get is always better than watch but I agree with having it here.